### ISSUE TITLE
drop-ins in /usr/lib/systemd/system take precendence over those in /etc/systemd/system

### ISSUE THREAD

**Issue by mbunkus** [systemd/systemd#13198]
**systemd version the issue has been seen with**

```
systemd 242 (242.84-1-arch)
+PAM +AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
```

**Used distribution**

Arch Linux, updated right before filing the bug

**Expected behaviour & erroneous behaviour**

On my installation there are drop-ins for the same unit in both `/usr/lib/systemd/system` and `/etc/systemd/system`. Both drop-ins touch the same settings, e.g. `User` for a service unit. I expected the setting I made in `/etc/systemd/system` to take effect, not the one from the drop-in in `/usr/lib/systemd/system`.

What actually happens, both according to simply starting the service & looking at it with `systemctl cat …`, is that the settings from `/usr/lib/systemd/system` drop-ins take effect.

Here's how to reproduce:

Create a unit & a drop-in for it in `/usr/lib/systemd/system` and a drop-in in `/etc/systemd/system`, have all three files set e.g. `User` to different values, reload `systemd` and look at the interpreted unit:

```
[0 mosu@sweet-chili ~] cat /usr/lib/systemd/system/dummy1.service
[Unit]
Description=Test Unit 1

[Service]
User=usrlib
[0 mosu@sweet-chili ~] cat /usr/lib/systemd/system/dummy1.service.d/user.conf
[Service]
User=usrlib.d
[0 mosu@sweet-chili ~] cat /etc/systemd/system/dummy1.service.d/override.conf
[Service]
User=etc.d
[0 mosu@sweet-chili ~] sudo systemctl daemon-reload
[0 mosu@sweet-chili ~] sudo systemctl cat dummy1.service
# /usr/lib/systemd/system/dummy1.service
[Unit]
Description=Test Unit 1

[Service]
User=usrlib

# /etc/systemd/system/dummy1.service.d/override.conf
[Service]
User=etc.d

# /usr/lib/systemd/system/dummy1.service.d/user.conf
[Service]
User=usrlib.d
```

This is completely surprising. Not only does the `systemd.unit` man page list the path load order, and that one clearly has `/etc/systemd/system` take precedence over `/usr/lib/systemd/system`, it also states explicitly which drop-ins take precedence over which other drop-ins. Here's what the man page says:

> In addition to /etc/systemd/system, the drop-in ".d/" directories for
> system services can be placed in /usr/lib/systemd/system or
> /run/systemd/system directories. Drop-in files in /etc take precedence
> over those in /run which in turn take precedence over those in
> /usr/lib.

This only works if the drop-ins have the same name in both cases. If I rename `/etc/systemd/system/dummy1.service.d/override.conf` to `/etc/systemd/system/dummy1.service.d/user.conf`, the drop-in from `/usr/lib` is suddenly not listed anymore. Another workaround is to name my custom drop-in something that sorts higher than `user.conf`, e.g. `zzz.conf`:

```
[0 mosu@sweet-chili ~] cd /etc/systemd/system/dummy1.service.d/
[0 mosu@sweet-chili /etc/systemd/system/dummy1.service.d] ls
override.conf
[0 mosu@sweet-chili /etc/systemd/system/dummy1.service.d] sudo mv override.conf user.conf
[0 mosu@sweet-chili /etc/systemd/system/dummy1.service.d] sudo systemctl daemon-reload
[0 mosu@sweet-chili /etc/systemd/system/dummy1.service.d] sudo systemctl cat dummy1.service
# /usr/lib/systemd/system/dummy1.service
[Unit]
Description=Test Unit 1

[Service]
User=usrlib

# /etc/systemd/system/dummy1.service.d/user.conf
[Service]
User=etc.d
[0 mosu@sweet-chili /etc/systemd/system/dummy1.service.d] sudo mv user.conf zzz.conf
[0 mosu@sweet-chili /etc/systemd/system/dummy1.service.d] sudo systemctl daemon-reload
[0 mosu@sweet-chili /etc/systemd/system/dummy1.service.d] sudo systemctl cat dummy1.service
# /usr/lib/systemd/system/dummy1.service
[Unit]
Description=Test Unit 1

[Service]
User=usrlib

# /usr/lib/systemd/system/dummy1.service.d/user.conf
[Service]
User=usrlib.d

# /etc/systemd/system/dummy1.service.d/zzz.conf
[Service]
User=etc.d
```

This is completely surprising to the user as `systemctl edit…` is suddenly not enough to ensure that your custom settings actually get applied. In fact, apart from choosing insane names for my drop-ins in `/etc/…` that always sort higher alphabetically there's nothing I, as an administrator, can do to ensure my own settings take precedence over the package maintainer's.

Please change the behavior to parse all drop-ins from `/usr/lib` first, then all from `/run` and last all from `/etc`. At the moment the drop-ins from all three locations seem to be put into the list, then sorted alphabetically, then parsed. Thanks.

Note: this is from the current `deluge` package on Arch Linux which follows such a setup, overwriting the user to run the daemon as in `/usr/lib/systemd/system/deluge-web.service.d/user.conf`. My guess is that the packager uses upstream's original unit file in `/usr/lib/systemd/system` and Arch's customization in the drop-in — but it prevents me from choosing my own user (safe from using my own, differently-named unit files).

**Comment by keszybz** [comment]
The paragraph you quoted continues:

> Multiple drop-ip files with different names are applied in lexicographic order, regardless of which of the directories they reside in.

This behaviour is on purpose and is not going to change. So there is no bug here.

That said, I agree that the docs are not very clear. That text has been amended and extended over the years, and it would be nice to rewrite it for clarity. A patch doing that would be very welcome.

**Comment by mbunkus** [comment]
Fair enough. Can you please point me to the rationale explaining what this is supposed to achieve? I'll have to include this idiosyncrasy when teaching systemd, and I'd like to know what to answer when folks ask me for why.

**Comment by keszybz** [comment]
It it possible to split the drop-ins into thematic areas, and allow overrides to any of the drop-ins separately. For unit files, this is of limited use (except for some possibly very complicated services), but it's used quite often for tmpfiles or udev rules. The same principles are used for all config file types, which allows users to carry expectations across different systemd services.

**Comment by mbunkus** [comment]
Thanks.

> which allows users to carry expectations across different systemd services.

The thing with expectations is that there are several point of views and several expectations here. One of them is that using `systemctl edit <unit>` will actually have an effect, which is enforced by the name chosen by `systemctl` for the drop-in: `override.conf`. The systemd project actively encourages this expectation, even though it is subtly wrong. Using `systemctl edit` always requires a cognitive load of having to check `systemctl cat <unit>` afterwards because `edit` doesn't guarantee that your settings take precedence.

I'm not trying to rant — I'm just trying to present a different perspective on the same issue. If you ever seriously think about changing this behavior, please keep this in mind. "Settings in `/etc` override `/run` override `/usr/lib`" was easy to each, easy to remember, easy to use.

**Comment by keszybz** [comment]
Yeah, the name "override" is not the best in some circumstances. Nevertheless, it works mostly OK for the most common case — when there are no dropins. I guess it'd be best to enhance `systemctl edit` to provide more information about dropins. We had some ideas to e.g. provide the existing definition of the unit in a comment, à la `git commit` and other git commands, that'd then be excluded from the file that is actually written to disk. That'd help your case, and would also make it much easier to write correct drop-ins.

**Comment by jjakob** [comment]
I'd argue that this is still an unsolved issue and needs to be reopened. I just ran into this when trying to override a package-supplied drop-in.
On Ubuntu, systemd-timesyncd is disabled when ntpd is installed with a drop-in `/lib/systemd/system/systemd-timesyncd.service.d/disable-with-time-daemon.conf` which contains `ConditionFileIsExecutable=!/usr/sbin/ntpd`

It's not possible for me to uninstall ntpd as it's a dependency of a different package that I don't want to uninstall. I've disabled ntp service but this obviously isn't enough.

To override this condition I created a drop-in `/etc/systemd/system/systemd-timesyncd.service.d/disable-condition-ntpd.conf` containing `ConditionFileIsExecutable=`. This didn't work and the condition was still tested regardless.

I've spent at least an hour trying to find the reason for this, finally finding this issue. Yes, I'd argue the lexical sorting is confusing at the very least, and very likely bad design. The documentation is unclear and there is no warning. Something like this would require a bold warning in the man page. It isn't logical to me that a drop-in in /etc/systemd/system wouldn't override a system provided drop-in. What even is the use of that? VERY bad design.
